src/pybind/mgr/dashboard/controllers/osd.py

```diff
@@ -229,6 +232,33 @@ def safe_to_destroy(self, ids):
                 'is_safe_to_destroy': False,
             }
 
+    @Endpoint('GET', query_params=['ids'])
+    @ReadPermission
+    def safe_to_remove(self, ids):
+        """
+        :type ids: int|[int]
+        """
+        # Unused now, might be useful for individual OSD checks in the future
+        _ = ids
+
+        # Check if there are any OSDs exceeds nearfull or full threshold.
+        all_health = health.HealthData(self._has_permissions, minimal=True).all_health()
+        checks = {c['type'] for c in all_health['health']['checks']}
+        unsafe_checks = set(['OSD_FULL', 'OSD_BACKFILLFULL', 'OSD_NEARFULL'])
+        failed_checks = checks & unsafe_checks
+        return {
+            'is_safe_to_remove': not bool(failed_checks),
+            'message': 'Removing OSD(s) is not recommended with these health checks: {}.'.
+                       format(', '.join(failed_checks)) if failed_checks else ''
+        }
+
+    @Endpoint('POST')
+    @DeletePermission
+    @handle_orchestrator_error('osd')
+    def remove_osds(self, osd_ids):
+        orch = OrchClient.instance()
+        orch.osds.remove(osd_ids)
```

Comment:
This one makes me nervous. We are putting all our safety-nets in the front-end, but the back-end (which we mean to keep as the official REST API) remains as a high-speed pass-through to the most dangerous commands in the CLI arena. And I'm afraid that every layer beneath the front-end will strictly follow the chain of command.

Reply:
How about adding a force flag and checks before passing to orchestrator?
* with `force off`: do not pass the request to orchestrator if checks are failed.
* with `force on`: pass the request to orchestrator. This can be used by a request from frontend that ignores the check warning.

Reply:
the cli and this endpoint need to run through the same checks. Adding any checks here won't help the cli user.

Reply:
Yeah, the thing is that most REST APIs don't usually allow to delete multiple resources in a single request, and they of course would use `DELETE /api/osd/<id>` for that (we are quite liberal on our REST API implementation by using `POST` to delete a resource).

Some APIs are using extra measures for this (e.g.: [S3 Multi-Factor Authentication Deletes](https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMFADelete.html)). Others, in fact, don't allow deletions (e.g.: s3 buckets) if the resource is not empty (or a force flag is required).

Probably we should also segregate DELETE permissions from Dashboard default roles (pool-manager, cluster-manager, block-manager, ...). That's for example, what GitHub REST API does for deleting a repo ([specific `delete_repo` scope is required](https://developer.github.com/v3/repos/#delete-a-repository)).

Another option (if we don't want to make development tasks/testing harder) would be to add a cluster-wide switch for Development/Production mode, so some security checks are enforced only in Production clusters.

Reply:
> Another option (if we don't want to make development tasks/testing harder) would be to add a cluster-wide switch for Development/Production mode, so some security checks are enforced only in Production clusters.

I am afraid that such tests will not be tested in detail if they are only active in a productive environment and not during development.